<a href="https://colab.research.google.com/github/KendallScott/QTW/blob/main/Case%20Study%205/Case_Study_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import label_binarize, OneHotEncoder, StandardScaler
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score, classification_report
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold, StratifiedKFold, train_test_split
from yellowbrick.classifier import ConfusionMatrix
from sklearn.preprocessing import StandardScaler
from numpy import arange
from sklearn.model_selection import GridSearchCV

#from example SVM notebook
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score
from sklearn.metrics import ConfusionMatrixDisplay

Prediction on the Action column, for use with a firewall.

Port is like an address that allows connection like a tv channel.


In [4]:
df = pd.read_csv("https://raw.githubusercontent.com/KendallScott/QTW/main/Case%20Study%205/log2.csv")
df

Source Port  Destination Port  NAT Source Port  NAT Destination Port  \
0            57222                53            54587                    53   
1            56258              3389            56258                  3389   
2             6881             50321            43265                 50321   
3            50553              3389            50553                  3389   
4            50002               443            45848                   443   
...            ...               ...              ...                   ...   
65527        63691                80            13237                    80   
65528        50964                80            13485                    80   
65529        54871               445                0                     0   
65530        54870               445                0                     0   
65531        54867               445                0                     0   

      Action    Bytes  Bytes Sent  Bytes Received  Packets  \
0      allow      177          94              83        2   
1      allow     4768        1600            3168       19   
2      allow      238         118             120        2   
3      allow     3327        1438            1889       15   
4      allow    25358        6778           18580       31   
...      ...      ...         ...             ...      ...   
65527  allow      314         192             122        6   
65528  allow  4680740       67312         4613428     4675   
65529   drop       70          70               0        1   
65530   drop       70          70               0        1   
65531   drop       70          70               0        1   

       Elapsed Time (sec)  pkts_sent  pkts_received  
0                      30          1              1  
1                      17         10              9  
2                    1199          1              1  
3                      17          8              7  
4                      16         13             18  
...                   ...        ...            ...  
65527                  15          4              2  
65528                  77        985           3690  
65529                   0          1              0  
65530                   0          1              0  
65531                   0          1              0  

[65532 rows x 12 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65532 entries, 0 to 65531
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Source Port           65532 non-null  int64 
 1   Destination Port      65532 non-null  int64 
 2   NAT Source Port       65532 non-null  int64 
 3   NAT Destination Port  65532 non-null  int64 
 4   Action                65532 non-null  object
 5   Bytes                 65532 non-null  int64 
 6   Bytes Sent            65532 non-null  int64 
 7   Bytes Received        65532 non-null  int64 
 8   Packets               65532 non-null  int64 
 9   Elapsed Time (sec)    65532 non-null  int64 
 10  pkts_sent             65532 non-null  int64 
 11  pkts_received         65532 non-null  int64 
dtypes: int64(11), object(1)
memory usage: 6.0+ MB


In [6]:
df.head()

Source Port  Destination Port  NAT Source Port  NAT Destination Port  \
0        57222                53            54587                    53   
1        56258              3389            56258                  3389   
2         6881             50321            43265                 50321   
3        50553              3389            50553                  3389   
4        50002               443            45848                   443   

  Action  Bytes  Bytes Sent  Bytes Received  Packets  Elapsed Time (sec)  \
0  allow    177          94              83        2                  30   
1  allow   4768        1600            3168       19                  17   
2  allow    238         118             120        2                1199   
3  allow   3327        1438            1889       15                  17   
4  allow  25358        6778           18580       31                  16   

   pkts_sent  pkts_received  
0          1              1  
1         10              9  
2          1              1  
3          8              7  
4         13             18

In [7]:
df['Source Port'] = df['Source Port'].astype(str)
df['Destination Port'] = df['Destination Port'].astype(str)
df['NAT Source Port'] = df['NAT Source Port'].astype(str)
df['NAT Destination Port'] = df['NAT Destination Port'].astype(str)

In [8]:
df.describe()

Bytes    Bytes Sent  Bytes Received       Packets  \
count  6.553200e+04  6.553200e+04    6.553200e+04  6.553200e+04   
mean   9.712395e+04  2.238580e+04    7.473815e+04  1.028660e+02   
std    5.618439e+06  3.828139e+06    2.463208e+06  5.133002e+03   
min    6.000000e+01  6.000000e+01    0.000000e+00  1.000000e+00   
25%    6.600000e+01  6.600000e+01    0.000000e+00  1.000000e+00   
50%    1.680000e+02  9.000000e+01    7.900000e+01  2.000000e+00   
75%    7.522500e+02  2.100000e+02    4.490000e+02  6.000000e+00   
max    1.269359e+09  9.484772e+08    3.208818e+08  1.036116e+06   

       Elapsed Time (sec)      pkts_sent  pkts_received  
count        65532.000000   65532.000000   65532.000000  
mean            65.833577      41.399530      61.466505  
std            302.461762    3218.871288    2223.332271  
min              0.000000       1.000000       0.000000  
25%              0.000000       1.000000       0.000000  
50%             15.000000       1.000000       1.000000  
75%             30.000000       3.000000       2.000000  
max          10824.000000  747520.000000  327208.000000

In [10]:
df[['Source Port','Action']].groupby('Source Port').count()

Action
Source Port        
0               173
10000             3
10001             2
10002             2
10003             8
...             ...
993              19
9957              3
9968              1
9969              1
9997              1

[22724 rows x 1 columns]

There are 22,724 distinct Source Ports in this dataset.

In [11]:
# Make dataframe for model input
X = df.loc[:, df.columns != 'Action']

#one hot encoding categorical variables for model
cols = X.columns
num_cols = X._get_numeric_data().columns
cat_cols = list(set(cols) - set(num_cols))
cat_cols

#creating dataframe of categorical columns
cat_df = X[cat_cols]
cat_df = pd.get_dummies(cat_df, columns=cat_df.columns)

#creating dataframe of numeric columns
scaler = StandardScaler()
num_df = scaler.fit_transform(X[num_cols])

X=cat_df.join(pd.DataFrame(num_df,columns=X[num_cols].columns, index=X[num_cols].index))
X

Destination Port_0  Destination Port_1  Destination Port_10004  \
0                       0                   0                       0   
1                       0                   0                       0   
2                       0                   0                       0   
3                       0                   0                       0   
4                       0                   0                       0   
...                   ...                 ...                     ...   
65527                   0                   0                       0   
65528                   0                   0                       0   
65529                   0                   0                       0   
65530                   0                   0                       0   
65531                   0                   0                       0   

       Destination Port_10010  Destination Port_10012  Destination Port_10016  \
0                           0                       0                       0   
1                           0                       0                       0   
2                           0                       0                       0   
3                           0                       0                       0   
4                           0                       0                       0   
...                       ...                     ...                     ...   
65527                       0                       0                       0   
65528                       0                       0                       0   
65529                       0                       0                       0   
65530                       0                       0                       0   
65531                       0                       0                       0   

       Destination Port_10030  Destination Port_10086  Destination Port_10096  \
0                           0                       0                       0   
1                           0                       0                       0   
2                           0                       0                       0   
3                           0                       0                       0   
4                           0                       0                       0   
...                       ...                     ...                     ...   
65527                       0                       0                       0   
65528                       0                       0                       0   
65529                       0                       0                       0   
65530                       0                       0                       0   
65531                       0                       0                       0   

       Destination Port_10104  ...  NAT Destination Port_995  \
0                           0  ...                         0   
1                           0  ...                         0   
2                           0  ...                         0   
3                           0  ...                         0   
4                           0  ...                         0   
...                       ...  ...                       ...   
65527                       0  ...                         0   
65528                       0  ...                         0   
65529                       0  ...                         0   
65530                       0  ...                         0   
65531                       0  ...                         0   

       NAT Destination Port_9992  NAT Destination Port_9999     Bytes  \
0                              0                          0 -0.017255   
1                              0                          0 -0.016438   
2                              0                          0 -0.017244   
3                              0                          0 -0.016695   
4                              0                  

Checking columns that have a high number of 0s (an infrequent categorical column).

In [12]:
Source_Port = []
Source_Port_perc = []

for i in X.columns:
  count = X.loc[X[i]==0,i].count()
  if count > 0:
    print('{}:'.format(i),
          '{}%'.format(np.round(np.divide(count,len(df))*100,2)))
    Source_Port_perc.append(np.round(np.divide(count,len(df))*100,2))
    Source_Port.append(i)

Streaming output truncated to the last 5000 lines.
NAT Source Port_63672: 100.0%
NAT Source Port_63675: 100.0%
NAT Source Port_63676: 100.0%
NAT Source Port_63683: 100.0%
NAT Source Port_6369: 100.0%
NAT Source Port_63690: 100.0%
NAT Source Port_63691: 100.0%
NAT Source Port_63692: 100.0%
NAT Source Port_63697: 100.0%
NAT Source Port_63699: 100.0%
NAT Source Port_63703: 100.0%
NAT Source Port_63705: 100.0%
NAT Source Port_63708: 100.0%
NAT Source Port_63709: 100.0%
NAT Source Port_6371: 100.0%
NAT Source Port_63710: 100.0%
NAT Source Port_63713: 100.0%
NAT Source Port_63714: 100.0%
NAT Source Port_63715: 100.0%
NAT Source Port_63716: 100.0%
NAT Source Port_63717: 100.0%
NAT Source Port_63725: 100.0%
NAT Source Port_63729: 100.0%
NAT Source Port_63730: 100.0%
NAT Source Port_63733: 100.0%
NAT Source Port_63735: 100.0%
NAT Source Port_63738: 100.0%
NAT Source Port_63744: 100.0%
NAT Source Port_63745: 100.0%
NAT Source Port_63749: 100.0%
NAT Source Port_6375: 100.0%
NAT Source Port_63751:

In [13]:
Source_Port_df = pd.DataFrame({'Source_Port': list(Source_Port), 'Source_Port_perc': list(Source_Port_perc)}, columns=['Source_Port', 'Source_Port_perc'])
Source_Port_df = Source_Port_df.loc[Source_Port_df['Source_Port_perc'] > 99.5]
Source_Port_df
drop_array = Source_Port_df.Source_Port.unique()

Removing Categorical columns with less than 0.5% of variation for Source Port

In [24]:
drop_array

array(['Destination Port_1', 'Destination Port_10004',
       'Destination Port_10010', ..., 'NAT Destination Port_995',
       'NAT Destination Port_9992', 'NAT Destination Port_9999'],
      dtype=object)

In [23]:
len(drop_array)

57633

In [20]:
str(drop_array[0])

'Destination Port_1'

In [ ]:
i=1
while i < len(drop_array):
  X = X.drop([drop_array[i]], axis=1)
  print('dropped ', drop_array[i] )
  i+=1

dropped  Destination Port_10004
dropped  Destination Port_10010
dropped  Destination Port_10012
dropped  Destination Port_10016
dropped  Destination Port_10030
dropped  Destination Port_10086
dropped  Destination Port_10096
dropped  Destination Port_10104


In [22]:
X.describe()

Destination Port_0  Destination Port_10004  Destination Port_10010  \
0                       0                       0                       0   
1                       0                       0                       0   
2                       0                       0                       0   
3                       0                       0                       0   
4                       0                       0                       0   
...                   ...                     ...                     ...   
65527                   0                       0                       0   
65528                   0                       0                       0   
65529                   0                       0                       0   
65530                   0                       0                       0   
65531                   0                       0                       0   

       Destination Port_10012  Destination Port_10016  Destination Port_10030  \
0                           0                       0                       0   
1                           0                       0                       0   
2                           0                       0                       0   
3                           0                       0                       0   
4                           0                       0                       0   
...                       ...                     ...                     ...   
65527                       0                       0                       0   
65528                       0                       0                       0   
65529                       0                       0                       0   
65530                       0                       0                       0   
65531                       0                       0                       0   

       Destination Port_10086  Destination Port_10096  Destination Port_10104  \
0                           0                       0                       0   
1                           0                       0                       0   
2                           0                       0                       0   
3                           0                       0                       0   
4                           0                       0                       0   
...                       ...                     ...                     ...   
65527                       0                       0                       0   
65528                       0                       0                       0   
65529                       0                       0                       0   
65530                       0                       0                       0   
65531                       0                       0                       0   

       Destination Port_10115  ...  NAT Destination Port_995  \
0                           0  ...                         0   
1                           0  ...                         0   
2                           0  ...                         0   
3                           0  ...                         0   
4                           0  ...                         0   
...                       ...  ...                       ...   
65527                       0  ...                         0   
65528                       0  ...                         0   
65529                       0  ...                         0   
65530                       0  ...                         0   
65531                       0  ...                         0   

       NAT Destination Port_9992  NAT Destination Port_9999     Bytes  \
0                              0                          0 -0.017255   
1                              0                          0 -0.016438   
2                              0                          0 -0.017244   
3                              0                          0 -0.016695   
4 

In [21]:
X = X.drop(str(drop_array[0]), axis=1)

In [2]:
X

NameError: ignored

NameError: ignored

In [ ]:
X.describe()

In [24]:
#creating dataframe to find where there is high readmitted rates for a variable
y_alt = df[['Action']]
y_alt = pd.get_dummies(y_alt, columns=y_alt.columns)

In [ ]:
model = LinearSVC(max_iter=100000,tol=0.01)
X_train,X_test, y_train,y_test = train_test_split(X,y_alt)

In [25]:
y_alt

Action_allow  Action_deny  Action_drop  Action_reset-both
0                 1            0            0                  0
1                 1            0            0                  0
2                 1            0            0                  0
3                 1            0            0                  0
4                 1            0            0                  0
...             ...          ...          ...                ...
65527             1            0            0                  0
65528             1            0            0                  0
65529             0            0            1                  0
65530             0            0            1                  0
65531             0            0            1                  0

[65532 rows x 4 columns]